### PR TITLE
Initial commit to get GLUE (BERT) on TPU

### PR DESCRIPTION
**Pull request by jysohn23**
Branched from: https://github.com/huggingface/transformers/blob/master/examples/run_glue.py.

1 score and accuracy looks pretty much identical to GPU when global batch size is the same.

Sample command:
```
python run_glue_tpu.py \
   --model_type bert \
   --model_name_or_path bert-base-cased \
   --task_name MRPC \
   --do_train \
   --do_eval \
   --do_lower_case \
   --data_dir ~/datasets/glue/MRPC \
   --max_seq_length 128 \
   --train_batch_size 32 \
   --learning_rate 3e-5 \
   --num_train_epochs 3.0 \
   --output_dir /tmp/MRPC \
   --overwrite_output_dir \
   --logging_steps 5 \
   --save_steps 50 \
   --num_cores=8 \
   --only_log_master
```

Eval metrics and train loss curves:
Screen Shot 2019-11-15 at 4 31 51 PM
Screen Shot 2019-11-15 at 4 31 57 PM

### CODE REVIEW

File: examples/run_glue_tpu.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+            model = model_class.from_pretrained(checkpoint)
+            model.to(args.device)
+            result = evaluate(args, model, tokenizer, prefix=prefix, disable_logging=disable_logging)
+            result = dict((k + '_{}'.format(global_step), v) for k, v in result.items())
+            results.update(result)
+
+    return results
+
+
+def get_args():
```
Comment (taylanbil) [inline examples/run_glue_tpu.py:411]: I assume the majority of args below are from the non-tpu version. Would it make sense to import that function and only add the tpu specific flags here in this file?

File: examples/run_glue_tpu.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+def set_seed(args):
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+
+
+def get_sampler(dataset):
+    if xm.xrt_world_size() <= 1:
+        return RandomSampler(dataset)
+    return DistributedSampler(dataset,
```
Comment (taylanbil) [inline examples/run_glue_tpu.py]: Is this style compatible w/ HF repo?

File: examples/run_glue_tpu.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+        for step, batch in enumerate(epoch_iterator):
+            # Save model checkpoint.
+            if args.save_steps > 0 and global_step % args.save_steps == 0:
+                output_dir = os.path.join(args.output_dir, 'checkpoint-{}-xla{}'.format(global_step, xm.get_ordinal()))
+                logger.info("Saving model checkpoint to %s", output_dir)
+                if not os.path.exists(output_dir):
+                    os.makedirs(output_dir)
+                model_to_save = model.module if hasattr(model, 'module') else model  # Take care of distributed/parallel training
+                model_to_save.save_pretrained(output_dir, xla_device=True)
+                torch.save(args, os.path.join(output_dir, 'training_args.bin'))
```
Comment (taylanbil) [inline examples/run_glue_tpu.py:140]: is this saving only from master ordinal?

File: examples/run_glue_tpu.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+                      'attention_mask': batch[1],
+                      'labels':         batch[3]}
+            if args.model_type != 'distilbert':
+                # XLM, DistilBERT and RoBERTa don't use segment_ids
+                inputs['token_type_ids'] = batch[2] if args.model_type in ['bert', 'xlnet'] else None
+            outputs = model(**inputs)
+            loss = outputs[0]  # model outputs are always tuple in transformers (see doc)
+
+            if args.gradient_accumulation_steps > 1:
+                loss = loss / args.gradient_accumulation_steps
```
Comment (taylanbil) [inline examples/run_glue_tpu.py:154]: We may want to mark step here, otherwise graph can grow too big. We received an issue in the fairseq repo related to this.

File: examples/run_glue_tpu.py [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+    # Training
+    if args.do_train:
+        train_dataset = load_and_cache_examples(args, args.task_name, tokenizer, evaluate=False)
+        global_step, tr_loss = train(args, train_dataset, model, tokenizer, disable_logging=disable_logging)
+        logger.info(" global_step = %s, average loss = %s", global_step, tr_loss)
+
+
+    # Saving best-practices: if you use defaults names for the model, you can reload it using from_pretrained()
+    output_dir = os.path.join(args.output_dir, 'final-xla{}'.format(xm.get_ordinal()))
+    if args.do_train:
```
Comment (taylanbil) [inline examples/run_glue_tpu.py]: line 354 is the same if. join?

File: examples/run_glue_tpu.py [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+def load_and_cache_examples(args, task, tokenizer, evaluate=False):
+    processor = processors[task]()
+    output_mode = output_modes[task]
+    # Load data features from cache or dataset file
+    cached_features_file = os.path.join(args.data_dir, 'cached_{}_{}_{}_{}_xla{}'.format(
+        'dev' if evaluate else 'train',
+        list(filter(None, args.model_name_or_path.split('/'))).pop(),
+        str(args.max_seq_length),
+        str(task),
+        xm.get_ordinal()))
```
Comment (taylanbil) [inline examples/run_glue_tpu.py:270]: this seems to be the only place in data loading that is world-aware. IIUC we're assuming the existence of filenames to shard the data, is that right?

File: examples/run_glue_tpu.py (hunk 2 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py]: Sorta. I've changed the linebreak to start with first arg.

File: examples/run_glue_tpu.py (hunk 3 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py:140]: No it's saving from each of the ordinals as `output_dir` is namespaced with ordinal#. In `xm.save` we set `master_only=False`.

File: examples/run_glue_tpu.py (hunk 4 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py:154]: Ah great point. Marked the step with a comment on why we're doing it.

File: examples/run_glue_tpu.py (hunk 6 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py:270]: Yes that's correct that we're loading with per core saving aware.

File: examples/run_glue_tpu.py (hunk 5 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py]: Good point, merged.

File: examples/run_glue_tpu.py (hunk 1 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py:411]: Not sure what you mean by "import that function". The way they had it in the other GLUE task runner looks like this: https://github.com/huggingface/transformers/blob/master/examples/run_glue.py#L334-L418. All other flags here will be used by this TPU runner script.

**Review by jysohn23 — COMMENTED** [review]
Also added a section on the README for running on TPU.

File: examples/run_glue_tpu.py (hunk 1 above)
Comment (taylanbil) [reply examples/run_glue_tpu.py:411]: Ah I see. If they had `get_args` function as you have here, you could import that function and use it here to avoid duplication. You can maybe upstream that change later. for now it's fine.

**Review by taylanbil — APPROVED** [review]

File: examples/run_glue_tpu.py (hunk 1 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py:411]: Ack.
